File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []

        memo = {
            "[" : "]",
            "{" : "}",
            "(" : ")"
            }

        for ele in s:
            if not stack or ele in memo:
                stack.append(ele)
            else:
                if memo.get(stack[-1], "-1") != ele:
                    return False
                stack.pop()

        return not stack
        

File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {

        Deque<Character> stack = new ArrayDeque<>();
        HashMap<Character, Character> mapp = new HashMap<>();
        mapp.put('(', ')');
        mapp.put('[', ']');
        mapp.put('{', '}');

        for(Character ch: s.toCharArray()) {
            if (ch == '(' || ch == '[' || ch == '{') {
                stack.push(ch);
            } else if (ch == ')' || ch == ']' || ch == '}') {
                if (stack.isEmpty()) {
                    return false;
                }
                Character open = stack.pop();
                if (mapp.get(open) != ch) {
                    return false;
                }
            }
        }

        return stack.isEmpty();
        
    }
}